[team-01][BE] 프로젝트 초기 세팅, API - 전체 기획전 데이터, 특정 반찬 상세 데이터 기능

donggi-lee-bit opened the pull request:
안녕하세요 `sigrid`, `donggi` 와 `jun` 입니다!! 프로젝트의 초기 세팅과 전체 기획전 데이터, 특정 반찬 상세 데이터 API 구현이 완료되어 리뷰 요청드립니다. PR 상세 메시지를 더 추가할 예정입니다. 2주간 잘 부탁드립니다, 감사합니다 😆

<br/><br/>

## 프로젝트 중점사항 및 구현내용  

첫 주차기 때문에 협업을 빨리 하기 위한 공통 룰, 공통 기능을 중점으로 프로젝트를 진행했습니다.

- [x] 백엔드 끼리의 원활한 협업을 위한 공통 예외처리 
- [x] 클라이언트측에 데이터를 빨리 내려주기 위한 필요한 API 기능 개발
- [x] 테이블 설계 및 ERD 작성

<br/><br/>

## 공통  

<details>
<summary>🌱 깃 관리</summary>
<div markdown="1">
<br/>

## 깃 브랜치
Backend에서는 feature/jun, feature/donggi, dev-BE 세 개의 브랜치를 통해 각자가 작업하고 dev-BE에 통합하는 방식으로 진행했습니다.

![fe-background](https://user-images.githubusercontent.com/92818747/164193085-7e16215a-00c9-47a7-9356-a3b6d8d357da.png)



- team-01: 마스터 브랜치 & 배포
- dev: BE, FE 머지한 공통 브랜치
- dev-BE: BE 디폴트 브랜치
- dev-FE: FE 디폴트 브랜치
- feature: 기능개발 브랜치
- PR마다 태그를 달아서 버전 명시하기

<br/><br/>

## 커밋 메시지 

커밋 메시지는 협의를 통해 아래와 같이 작성하였습니다.

| Type | Description |
| --- | --- |
| Init | 초기화 |
| Feat | 새로운 기능 추가 |
| Fix | 버그 수정 |
| Docs | 문서 작업, 수정 |
| Refactor | 코드 리팩토링 |
| Style | 코드 스타일 및 포맷 변경(코드 자체의 변경은 없는 경우, 함수/ 변수명 변경 포함) |
| Test | 테스트 코드 작성, 테스트 리팩토링(프로덕션 코드 변경 X) |
| Chore | 소스 코드를 건들지 않는 작업 - 빌드 테스트 업데이트, 패키지 매니저를 설정하는 경우(ex: gitignore) |
| Design | CSS 등 UI 디자인 변경 |
| Ci | CI 관련 설정 수정 |
| Rename | 파일 혹은 폴더명을 수정하거나 옮기는 작업만 수행 |
| Remove | 파일을 삭제하는 작업만 수행 |
| Comment | 필요한 주석 추가 및 변경 |
| Build | 빌드 관련 파일 수정 |
| !BREAKING CHANGE | 커다란 API 변경 |
| !HOTFIX | 급하게 치명적인 버그를 고쳐야 하는 경우 |

<br/><br/><br/>


## 깃 협업 툴 

[프로젝트](https://github.com/hemudi/sidedish/projects), [이슈](https://github.com/hemudi/sidedish/issues), [WIKI](https://github.com/hemudi/sidedish/wiki)

</div>
<br/><br/><br/>
</details>

<br/><br/>

## Backend

<details>
<summary>📌  API </summary>
<div markdown="1">
<br/>

현재 작성된 API는 세 개 정도 밖에 없는데 설계한 API에서 잘못된 부분은 없는지, 더 좋은 API가 있을지 알고 싶습니다.

<br/>

```
GET /api/dishes/{id}
```
<details>
<summary> 상품 상세 API </summary>
<div markdown="1">
<br/>


```
{
    "success": true,
    "response": {
        "dishId": 1,
        "name": "한식",
        "description": "한국 전통음식",
        "normalPrice": 39800,
        "salePrice": 32000,
        "badge": "NONE",
        "deliveryType": "NONE",
        "thumbnail": "www.naver.com",
        "dishStatus": "OUT_OT_STOCK",
        "categoryId": 1
    },
    "error": null
}
```
</div>
<br/><br/><br/>
</details>

<br/><br/>

```
GET /api/dishes
```
<details>
<summary> 상품 기획전 API </summary>
<div markdown="1">
<br/>


```
{
    "success": true,
    "response": [
        {
            "dishId": 1,
            "name": "한식",
            "description": "한국 전통음식",
            "normalPrice": 39800,
            "salePrice": 32000,
            "badge": "NONE",
            "deliveryType": "NONE",
            "thumbnail": "www.naver.com",
            "dishStatus": "OUT_OT_STOCK",
            "categoryId": 1,
            "title": "한식"
        },
        {
            "dishId": 2,
            "name": "한식",
            "description": "일본 전통음식",
            "normalPrice": 49800,
            "salePrice": 48000,
            "badge": "NONE",
            "deliveryType": "NONE",
            "thumbnail": "www.naver.com",
            "dishStatus": "OUT_OT_STOCK",
            "categoryId": 2,
            "title": "일식"
        },
        {
            "dishId": 3,
            "name": "중식",
            "description": "중국 전통음식",
            "normalPrice": 38000,
            "salePrice": 36000,
            "badge": "NONE",
            "deliveryType": "NONE",
            "thumbnail": "www.yahoo.com",
            "dishStatus": "OUT_OT_STOCK",
            "categoryId": 3,
            "title": "중식"
        }
    ],
    "error": null
}
```
</div>
<br/><br/>
</details>

<br/><br/>

```
GET /api/dishes/categories
```
<details>
<summary> 상품 카테고리 별 조회  </summary>
<div markdown="1">
<br/>


```
{
    "success": true,
    "response": {
        "planningData": {
            "중식": [
                {
                    "dishId": 3,
                    "name": "중식",
                    "description": "중국 전통음식",
                    "normalPrice": 38000,
                    "salePrice": 36000,
                    "badge": "NONE",
                    "deliveryType": "NONE",
                    "thumbnail": "www.yahoo.com",
                    "dishStatus": "OUT_OT_STOCK",
                    "categoryId": 3,
                    "title": "중식"
                }
            ],
            "일식": [
                {
                    "dishId": 2,
                    "name": "한식",
                    "description": "일본 전통음식",
                    "normalPrice": 49800,
                    "salePrice": 48000,
                    "badge": "NONE",
                    "deliveryType": "NONE",
                    "thumbnail": "www.naver.com",
                    "dishStatus": "OUT_OT_STOCK",
                    "categoryId": 2,
                    "title": "일식"
                }
            ],
            "한식": [
                {
                    "dishId": 1,
                    "name": "한식",
                    "description": "한국 전통음식",
                    "normalPrice": 39800,
                    "salePrice": 32000,
                    "badge": "NONE",
                    "deliveryType": "NONE",
                    "thumbnail": "www.naver.com",
                    "dishStatus": "OUT_OT_STOCK",
                    "categoryId": 1,
                    "title": "한식"
                }
            ]
        }
    },
    "error": null
}
```
</div>
<br/><br/>
</details>



</div>
<br/><br/><br/>
</details>


<details>
<summary>💾 테이블 설계</summary>
<div markdown="3">
<br/>


![Untitled](https://user-images.githubusercontent.com/92818747/164193275-bd149116-f240-4010-95f2-ae89ff3c4f77.png)


## users

사용자 테이블

|컬럼|내용|특징|
|:------|:----|:---|
|&nbsp;&nbsp;user_id|&nbsp;&nbsp;식별자|PK|
|&nbsp;&nbsp;login_id|&nbsp;&nbsp;사용자 로그인 아이디|
|&nbsp;&nbsp;password|&nbsp;&nbsp;사용자 비밀번호|


<br/><br/>

## order

주문 테이블

|컬럼|내용|특징|
|:------|:----|:---|
|&nbsp;&nbsp;order_id|&nbsp;&nbsp;식별자|PK|
|&nbsp;&nbsp;order_price|&nbsp;&nbsp;총 주문 금액|
|&nbsp;&nbsp;dish_id|&nbsp;&nbsp;음식 테이블 식별자|FK|
|&nbsp;&nbsp;user_id|&nbsp;&nbsp;사용자 테이블 식별자|FK|

<br/><br/>

## orderdetail

주문상세 테이블

|컬럼|내용|특징|
|:------|:----|:---|
|&nbsp;&nbsp;orderdetail_id|&nbsp;&nbsp;식별자|PK|
|&nbsp;&nbsp;dish_count|&nbsp;&nbsp;상품 주문 수량|
|&nbsp;&nbsp;order_price|&nbsp;&nbsp;상품 주문 금액||
|&nbsp;&nbsp;dish_id|&nbsp;&nbsp;음식 테이블 식별자|FK|
|&nbsp;&nbsp;order_id|&nbsp;&nbsp;주문 테이블 식별자|FK|


<br/><br/>

## category
카테고리 테이블

|컬럼|내용|특징|
|:------|:----|:---|
|&nbsp;&nbsp;category_id|&nbsp;&nbsp;식별자|PK|
|&nbsp;&nbsp;title|&nbsp;&nbsp; 카테고리 이름|

<br/>

## images
이미지 테이블

|컬럼|내용|특징|
|:------|:----|:---|
|&nbsp;&nbsp;image_id|&nbsp;&nbsp;식별자|PK|
|&nbsp;&nbsp;image_url|&nbsp;&nbsp; 이미지 주소|
|&nbsp;&nbsp;dish_id|&nbsp;&nbsp;음식 테이블 식별자|FK|

</div>
<br/><br/><br/>
</details>



<br/><br/>

## 질문사항

<details>
<summary>❓ 깃 질문사항</summary>
<div markdown="1">
<br/>


프로젝트를 진행하며 의문이 들었던 부분을 작성한 탭입니다.

### 1. 중복된 컬럼 - Donggi
- 반찬 테이블은 원 가격, 할인된 가격, 할인에 대한 뱃지를 가지고 있습니다.
- 클라이언트 측에서 원 가격과 할인된 가격, 뱃지를 전달해달라는 요청이 왔습니다.
- 저희 생각은 원 가격과 할인된 가격 두 컬럼을 다 가지고 있는 것은 중복된다고 판단했습니다.
  이 과정에서 두 컬럼 모두 가지고 있는게 좋을지, 하나의 가격 컬럼과 할인율만 가져가고 계산은 클라이언트측에서 하는 것이 좋을지 리뷰어님의 의견을 듣고싶습니다.

<br/><br/>

### 2. 조인 - Jun

음식과 카테고리 테이블을 조회하게 되면 음식 하나에 여러개의 카테고리가 나오게 됩니다. 이렇게 되면 컬럼이 뻥튀기(?) 되는 현상이 발생하게 되는데 조인 방향이 옳은지, 아래 쿼리문에서 수정해야 할 사항이 있는지
리뷰어님께 PR받고 싶습니다.

````java
public List<PlanningDataRequest> getPlanningData() {
    String query = "SELECT dish.*, category.* FROM dish JOIN category ON dish.category_id = category.category_id";
    return namedParameterJdbcTemplate.query(query, generalMapper);
}
````

```text
빠르게 데이터를 내려주기 위해 *를 사용했는데 추후 모든 컬럼을 명시해주겠습니다.
```

<br/><br/>

### 3. 테이블 설계 - Donggi, Jun

- 테이블 전반에 대해 방향이 올바른지, 크게 어긋난 부분이 없는지 알고 싶습니다.
- 카테고리 밑에 서브 카테고리를 둬서 연관관계를 맺고 싶은데 이때 서브 카테고리와 상품의 연관관계가 어떻게 맺어질지 감이 잘 오지 않습니다.

```text
상품 - 카테고리(다대 일)
서브 카테고리 - 카테고리 (다대 일)
카테고리 - 상품(다대 일) - 이 부분입니다.
```


</div>
<br/><br/><br/>
</details>

<br/><br/><br/>

devjun10 committed:
Chore: 프로젝트 다시 세팅

- 프로젝트 설정 초기화

devjun10 committed:
Feat: 의존성 추가 #1

- 의존성 추가 및 세팅 초기화

devjun10 committed:
Feat: 공통 예외처리 추가

- ExceptionHandler 추가
- 예외 등록

devjun10 committed:
Feat: 도메인 클래스 개발

- Category, Dish 도메인 클래스 설계

devjun10 committed:
Feat: 기획전 데이터 API 설계

- 컨트롤러, 서비스, 레포지토리 연결

devjun10 committed:
Feat: 조회용 API 연결

- 프론트에 데이터를 내려주기 위한 API 개발
- 레포지토리 매핑

devjun10 committed:
Chore: 폴더 정리

- 컨트롤러, 서비스, 레포지토리 및 디렉토리 정리

devjun10 committed:
Feat: 조회 API #5

- API 응답 클래스 생성
- 조회 API 설계

devjun10 committed:
Feat: 조회 API 기능 추가

- 카테고리 별 API 조회 추가

donggi-lee-bit committed:
Feat: API 명세 확인을 위한 swagger 추가 - #12

- build.gradle 에 의존성 추가

donggi-lee-bit committed:
Merge pull request #13 from hemudi/feature/api-docs

Feat: API 명세 확인을 위한 swagger 추가 - #12

sigridjineth commented:
안녕하세요, 리뷰어 시그리드입니다.
먼저 2주간 잘 부탁드리구요. 언제든지 궁금하신 점이 있으면 메시지나 PR을 자주 남겨주시면 디스커션과 저의 팔로업에도 도움이 됩니다.

sigridjineth commented:
### API 관련
1. 웹 프론트엔드 세계에서의 변수 이름짓는 컨벤션은 `camel_case` 이기 때문에, 그렇지 않게 명명된 변수의 이름을 바꿔주는 것이 좋지 않을까 생각합니다. 해당 부분은 프론트엔드 파트너 분들과 소통하여 해결하시기 바랍니다.
2. 별도의 `"NONE"` 메시지를 남겨주기 보다는, 없는 경우에는 그냥 빈 문자열을 던지는 것이 좋지 않을까라는 개인적인 생각이 있습니다.

sigridjineth commented:
깃 브랜치 관리 기법이 인상적이네요. 저 그림은 일일이 그리신 건가요?

sigridjineth commented:
### 스키마 관련
* 김영한 선생님의 강의를 보면 `일대다 단방향 매핑` 이 좋지 않다고 합니다. 위에 올려주신 스키마도 `일대다 양방향 매핑` 이 되어 있네요. 일대다 단방향이 아닌, 양방향으로 매핑 전략을 가져간 것에 대한 본인만의 주장과 근거를 제시해주세요. 지금 바로 완전한 답변을 주지 않으셔도 좋고, 점진적으로 학습 과정 속에서 느끼시는 사항을 답변으로 남겨주셔도 좋습니다.
* 우리의 현재 요구사항에서 `dish` 테이블이 `orderdetail` 테이블의 정보를 참조해야 하는 내용이 있나요?

### 테이블 설계 관련
* 테이블 설계에서 하나의 테이블을 두 개의 테이블로 분리하여 보관하려는 것 같은데요. 그렇게 한다면 이유는 무엇인가요? 만약 상품의 정보를 별도의 테이블로 빼서 관리한다고 하면, 해당 테이블을 별도로 뺀 것에 대한 정보를 한 번 상위 개념에서 묶어줄 수 있을 것이고, 그러면 1:1 매핑 관계가 성립할 수 있지 않을까요?

* 지금은 DB 스키마 설계를 먼저 했고 이에 바탕해서 데이터를 넣어보시면서 작업하려는 것 같은데요. 이것도 분명 훌륭한 방법인데, 저는 개인적으로 객체 설계를 먼저 하는 방향을 좋아합니다. 이는 철저히 개인 취향이니 만큼 한 번 고민해보시면 좋을 것 같습니다.

### 조인 관련
* 글쎄요. 구체적으로 어떻게 칼럼이 뻥튀기 된다고 말씀하시는 지 궁금하네요. `*` 표시로 내려주셔서 그런 것 아닐까요?
* 지금 `JOIN` 을 하고 계신데, `OUTER JOIN` 과 `INNER JOIN` 의 차이가 무엇인지 학습해보고 본인만의 언어로 기술해주세요.

realsigridjin commented on BE/src/main/java/com/example/be/config/SwaggerConfig.java:
```diff
@@ -0,0 +1,41 @@
+public class SwaggerConfig {
+
+	@Bean
+	public Docket apiV1() {
```
swagger 설정을 능수능란하게 사용하시는 군요. 멋집니다 👍

realsigridjin commented on BE/src/main/java/com/example/be/config/SwaggerConfig.java:
```diff
@@ -0,0 +1,41 @@
+			"코드스쿼드 2주차 팀 프로젝트 API 명세서입니다.",
+			"0.0.1",
+			"http://3.39.42.204/api/dishes/",
+			new Contact("Contact Me", "https://codesquad-members.slack.com/archives/C03BU6F4F5H", "devdonggilee@gmail.com"),
```
`nit`
해당 슬랙 링크는 외부에서는 접근하기 어려울 것이고, 추후 이 프로젝트를 찾아보시더라도 슬랙 멤버십의 한계상 아마 옛날 글을 보기 어려우실 겁니다. 링크 교체를 고민해보세요.

donggi-lee-bit replied:
Contact 아무도 하지 않겠지 생각하며 임의의 `url` 이라고 생각하고 했는데 추후에 확인하는 걸 고려하여 `url` 수정하도록 하겠습니다! 감사합니다!!

realsigridjin commented on BE/src/main/java/com/example/be/controller/dish/DishController.java:
```diff
@@ -0,0 +1,39 @@
+
+    @GetMapping
+    public ApiResult<List<PlanningDataRequest>> getPlanningData() {
+        return OK(dishService.getPlanningData());
```
ApiResult 를 깔끔하게 매핑해서 좋네요.
그런데 별도의 ApiResult를 굳이 만들 필요가 있나요? 만들더라도, 스프링 환경에서 자주 이용되는 ResponseEntity 를 활용해보는 것은 어떻게 생각하시나요?

[참고할 만한 링크](https://tecoble.techcourse.co.kr/post/2021-05-10-response-entity/)

devjun10 replied:
좋은 방법 같습니다. 이번에 다른 분들과 예외처리를 공부하면서 반환타입을 커스텀하는 것에 대해 이야기가 나와 만들어보게 되었습니다. 이렇게 객체를 만들어 사용하는 방법은 별로일까요? 🥺

realsigridjin replied:
아니요, 전 직접 만들어보며 구현하는 것이 아주 좋은 학습 태도라고 생각합니다. 다만, 바퀴를 굳이 재발명할 필요는 없으니까요 ㅎㅎ

realsigridjin commented on BE/src/main/java/com/example/be/controller/dish/DishDetail.java:
```diff
@@ -0,0 +1,74 @@
+    private Long categoryId;
+
+    public DishDetail(Dish dish) {
+        this.dishId = dish.getDishId();
```
DishDetail에 필요한 정보를 Dish로부터 인자를 받아서 직접 꺼내서 쓰는 것이 좋은 방향이라고 느껴지네요.

realsigridjin commented on BE/src/main/java/com/example/be/domain/dish/DeliveryType.java:
```diff
@@ -0,0 +1,5 @@
+package com.example.be.domain.dish;
+
+public enum DeliveryType {
```
`DeliveryType` 은 무엇을 위한 클래스 일까요?

devjun10 replied:
배송 타입을 넣어달라는 클라이언트 측의 요구사항을 맞추기 위해 넣었습니다.

realsigridjin commented on BE/src/main/java/com/example/be/domain/dish/Dish.java:
```diff
@@ -0,0 +1,89 @@
+    private String name;
+    private String description;
+    private BigDecimal normalPrice;
+    private BigDecimal salePrice;
```
학습 키워드: `BigDecimal` 과 `double` 의 차이는 무엇인가요?

donggi-lee-bit replied:
아래의 MySql 에서의 decimal 과 double 차이에 대한 질문을 주셔서 해당 답변 링크를 남기겠습니다! https://github.com/codesquad-members-2022/sidedish/pull/24#discussion_r855290047

realsigridjin commented on BE/src/main/java/com/example/be/domain/dish/DishStatus.java:
```diff
@@ -0,0 +1,6 @@
+
+public enum DishStatus {
+    OUT_OT_STOCK;
+
```
`IN_STOCK` 이 아직 없네요? ㅠㅠ

devjun10 replied:
ㅜㅜ

realsigridjin commented on BE/src/main/java/com/example/be/service/dish/DishService.java:
```diff
@@ -0,0 +1,36 @@
+        this.dishRepository = dishRepository;
+    }
+
+    @Transactional
```
이 메소드에서는 `(readOnly = true)` 옵션이 걸려있지 않은 이유가 무엇일까요?

realsigridjin replied:
`@Transactional` 에서 `(readOnly = true)` 옵션은 어떻게 동작하나요?

devjun10 replied:
readOnly를 빼먹은 것 같습니다. 다음 PR 때 수정해서 반영하겠습니다!

devjun10 replied:
`readonly = true` 부분이 동작하는 방법은 조금 더 학습이 필요할 것 같습니다. 다음 리뷰 때 이 부분을 추가 학습해서 코멘트를 달겠습니다!  PR 받는 과정에서 학습한 만큼만 설명해보겠습니다. 

<br/><br/>

1. `readonly = true` 설정을 사용하게 되면 Jdbc의 setReaonly(true)가 호출됩니다.
2.  스프링은 이 과정에서 쿼리에 힌트를 주고 Lock을 걸지 않습니다.
3.  FlushMode.NEVER와 같은 효과가 적용되며 수동으로 업데이트가 할 수도 없습니다.

<img width="638" alt="스크린샷 2022-04-22 오전 12 46 31" src="https://user-images.githubusercontent.com/92818747/164499773-877ef40b-5b84-49b1-9465-dd0847a8fd88.png">



> 이는 데이터베이스마다, 구현체마다 각 다른 특징을 가지고 있습니다. 


<br/><br/>

이를 통해 성능상 얻을 수 있는 이점은 아래와 같습니다. 
  - 트랜잭션은 각 ID가 부여되는데, 이 경우 별도의 트랜잭션 ID가 부여되지 않기 때문에 불필요하게 ID를 추가하는 작업이 없다.
  - 쿼리에 힌트를 주어 테이블과 컬럼에 Lock을 적용할 필요가 없다. 
  - 데이터의 일관성을 보장할 수 있다.  

<br/><br/>

> 추가 질문 사항이 있습니다. 

Jdbc의 JdbcTransactionObjectSupport에도 setReadonly(True)가 존재하고, TransactionSynchronizationManager에도 비슷한 setCurrentTransactionReadOnly  메서드가 존재하는데 어떤 것을, 어떤 순서로 사용하게 되는지가 궁금합니다.
<br/>

<img width="667" alt="스크린샷 2022-04-22 오전 1 02 57" src="https://user-images.githubusercontent.com/92818747/164503089-cedc32f6-0fc0-41ff-941f-1e3fca9d64ca.png">
<img width="673" alt="스크린샷 2022-04-22 오전 1 02 10" src="https://user-images.githubusercontent.com/92818747/164503107-88327de2-12bf-4853-a256-dbb1d2b5f034.png">

realsigridjin replied:
해당 옵션에 대한 설명도 넣어주세요!

devjun10 replied:
이 부분에 대해서 정확히 몰라서 조심스러운데 제가 학습한 만큼만 작성해보겠습니다.
<br/>

### setReadonly
Jdbc관련 setReadonly는 Jdbc 자체에서 데이터베이스에 거는 설정으로 쿼리에 힌트를 주고, 데이터베이스 차원에서 이를 지원하면 FlushMode.NEVER 모드로 변경되는 것입니다. 

<br/>

### isCurrentTransactionReadonly
하이버네이트의 flush 모드와 관련된 트랜잭션 설정입니다.

realsigridjin replied:
아래 링크들도 한 번 살펴보세요. 우리는 MySQL을 사용하고 있으니, 해당 구현체에 초점을 맞춰 학습할 필요가 있겠습니다.

[03. 당신의 @Transactional의 readOnly 옵션 잘 쓰고 있나요?](https://tech.yangs.kr/22)
[@Transactional(readOnly = true)는 필요한가?](https://cupeanimus.tistory.com/90)
[Spring Transaction by 우아한형제들 기술이사 권남님](https://kwonnam.pe.kr/wiki/springframework/transaction)
[Java 에서 DataBase Replication Master/Slave (write/read) 분기 처리하기](http://kwon37xi.egloos.com/5364167)

realsigridjin replied:
TransactionSynchronizationManager의 setCurrentTransactionReadOnly 동작 방식에 대해서는 하기 링크를 참조하세요.
[Spring Data JPA - Write, Read Only 분리 적용하기 - 1. 설정 및 원리](https://devs0n.tistory.com/60)

이런 것 보면 역시 스프링 계열은 한국어로 된 좋은 자료가 많다는 생각이 드네요. 워낙 학습자가 많아서 그렇네요.

짧게 언급해보자면, 우리가 `AbstractRoutingDataSource` 를 사용하지 않으면 다음과 같이 동작합니다.

* @Transactional이 걸린 메서드 호출
* Transaction 시작
   * DataSource로 부터 Connection 획득
   (TransactionSynchronizationManager.isCurrentTransactionReadOnly == false)
   * 해당 Transaction이 Read Only Transaction 인지 지정
   (TransactionSynchronizationManager.isCurrentTransactionReadOnly == true or false)
* 메서드 내부 코드 실행
* Transaction 종료

그런데 우리가 `AbstractRoutingDataSource` 를 활용하여 Data Connection Lazy Loading을 가능하도록 하면 다음과 같이 Transaction의 시작 순서를 바꾸어 `ReadOnly` 의 옵션을 `true` 로 만들 수 있습니다.

* @Transactional이 걸린 메서드 호출
* Transaction 시작
  * LazyConnectionDataSourceProxy로 부터 Connection Proxy 획득
  (TransactionSynchronizationManager.isCurrentTransactionReadOnly == false)
  * 해당 Transaction이 Read Only Transaction 인지 지
  (TransactionSynchronizationManager.isCurrentTransactionReadOnly == true or false)
* 메서드 내부 코드 실행
  * Connection을 통해 DB에 query를 보내야하므로 Connection 직접 사용 시도
  * AbstractRoutingDataSource로 부터 Connection 획
  (TransactionSynchronizationManager.isCurrentTransactionReadOnly == true or false)
  * Connection을 통해 DB에 query를 보냄
* Transaction 종료


위와 같이 내부적으로 DataSource로 부터 Connection을 얻는 과정을 LazyConnectionDataSourceProxy을 통해 역전을 시킬 수 있게 되므로 `@Transactional` 의 `readOnly` 속성에 따라 다른 DataSource를 사용하도록 지정할 수 있게 된 것입니다.

devjun10 replied:
이 부분에 대해 학습한 후 따로 정리 하겠습니다! 친절한, 상세한 답변 그리고 학습자료까지 진심으로 감사합니다 🙏

realsigridjin commented on BE/src/main/resources/application.yml:
```diff
@@ -0,0 +1,6 @@
+    url: jdbc:mysql://127.0.0.1:3306/test22
+    driver-class-name: com.mysql.cj.jdbc.Driver
+    username: root
+    password: 1234
```
`nit`
외부 환경변수로 주입하는 것을 생활화 합시다.

realsigridjin replied:
yml을 통해 DB를 붙인 것이 훌륭합니다 👍

devjun10 replied:
실수한 것 같습니다. 다음 리뷰 때 반영하겠습니다!

realsigridjin commented on BE/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,30 @@
+CREATE TABLE dish
+(
+    dish_id       BIGINT       NOT NULL PRIMARY KEY AUTO_INCREMENT,
+    name          VARCHAR(255) NULL,
```
이름을 255자나 지정할 이유는 없다고 생각해요.

donggi-lee-bit replied:
맞습니다..!!! 수정하도록 하겠습니다!

realsigridjin commented on BE/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,30 @@
+    dish_id       BIGINT       NOT NULL PRIMARY KEY AUTO_INCREMENT,
+    name          VARCHAR(255) NULL,
+    description   VARCHAR(255) NULL,
+    normal_price  DECIMAL      NULL,
```
MySQL에서 decimal과 double의 차이는 무엇인지 본인만의 언어로 정리해봅시다.
* 학습 키워드: 고정 소수점, 부동 소수점
[참고할 만한 링크](http://minsql.com/mysql/MySQL-decimal-type-%EC%A0%9C%EB%8C%80%EB%A1%9C-%EC%95%8C%EA%B3%A0-%EC%82%AC%EC%9A%A9%ED%95%98%EA%B8%B0/)

donggi-lee-bit replied:
알려주신 키워드를 바탕으로 MySQL 에서 `decimal` 과 `double` 의 차이에 대해서 학습해보았습니다.

## double

`double` 은 `float` 와 같이 MySQL 에서 실수를 저장할 수 있는 타입입니다. 두 타입은 부동 소수점을 가지며, 부동 소수점은 매우 큰 실수까지 표현할 수 있습니다.

## decimal

`decimal` 은 `int` 와 함께 고정 소수점을 가집니다. 고정 소수점은 부동 소수점과는 달리 표현할 수 있는 범위가 매우 적습니다.

## MySQL 에서의 decimal 과 double

MySQL 에서 실수를 표현할 때 `decimal` 과 `float`, `double` 을 사용할 수 있습니다. 부동 소수점을 가지는 `double` 과 `float` 을 사용하면 누적 연산이 많아졌을 땐 예상 값과 다른 결과 값이 나올 수 있기 때문에, 이런 경우에서는 `decimal` 을 사용하는 게 좋은 것 같습니다!

realsigridjin commented on BE/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,30 @@
+    normal_price  DECIMAL      NULL,
+    sale_price    DECIMAL      NULL,
+    badge         VARCHAR(255) NULL,
+    delivery_type VARCHAR(255) NULL,
```
MySQL에서 `enum` 타입 사용을 고려하지 않으시나요? 그렇다면, 그 이유는 무엇인가요?

donggi-lee-bit replied:
먼저 MySql 에서 `enum` 타입이 사용되고 있는 지 몰라서 고려하지 않았습니다 😂  언급해주셔서 MySql 의 `enum` 타입에 대해 간단하게 학습해보았습니다.

MySql 에서 `enum` 타입은 Java 에서 사용되는 `enum` 과 비슷하게 사용되고 있다고 느껴졌습니다. 하지만 `enum` 을 사용함에 있어 생기는 비용, 재사용성 등 생각보다 신중하게 써야한다고 느껴졌습니다! 해당 프로젝트에서는 그것까지 고려할 사항은 없으니 `enum` 타입을 사용해보도록 하겠습니다! 감사합니다 😆

realsigridjin commented on BE/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,30 @@
+    delivery_type VARCHAR(255) NULL,
+    thumbnail     VARCHAR(255) NULL,
+    category_id   BIGINT       NOT NULL,
+    dish_status   VARCHAR(255) NULL
```
`nit`

커멘트를 남길 수도 있답니다.

```suggestion
    dish_status   VARCHAR(255) NULL COMMENT '커멘트를 남길 수 있습니다. 이런 케이스가 있고 저런 케이스가 있습니다.',
```

donggi-lee-bit replied:
`COMMENT` 를 통해 어떠한 `column` 인지 추가하도록 하겠습니다. 감사합니다 ㅎㅎ

realsigridjin commented on BE/src/main/resources/schema.sql:
```diff
@@ -0,0 +1,30 @@
+    badge         VARCHAR(255) NULL,
+    delivery_type VARCHAR(255) NULL,
+    thumbnail     VARCHAR(255) NULL,
+    category_id   BIGINT       NOT NULL,
```
Foreign Key를 잡는 부분인 것 같은데, `constraints` 가 없는 이유가 무엇일까요?

donggi-lee-bit replied:
`constraint` 추가하였습니다 감사합니다 !!

realsigridjin commented on BE/src/main/java/com/example/be/repository/dish/DishReadRepository.java:
```diff
@@ -0,0 +1,41 @@
+
+    public static RowMapper<PlanningDataRequest> generalMapper = (rs, rowNum) ->
+            new PlanningDataRequest(
+                    rs.getLong("dish_id"),
```
`nit`
만약 이 상황에서 컬럼명이 변경된다면 어떻게 처리할 수 있을까요?

devjun10 replied:
상수로 빼서 처리하는 것이 좋을 것 같습니다.

devjun10 replied:
미션 중 요구사항이  `Spring Data Jdbc 사용` 인데 클라이언트측에 빠르게 데이터를 내려주기 위해 익숙한 내용을 사용했습니다. 다음 PR 때 삭제하겠습니다!

realsigridjin replied:
@devjun10 이번 요구사항이 Spring Data JDBC 사용이군요. 멋진 활약을 기대해보겠습니다 👍

realsigridjin commented on BE/src/main/java/com/example/be/service/dish/DishService.java:
```diff
@@ -0,0 +1,36 @@
+@Service
+public class DishService {
+
+    private final DishReadRepository dishReadRepository;
```
`ReadRepository` 를 별도로 나누신 이유가 무엇인지 궁금합니다.

devjun10 replied:
단순 조회와 엔티티의 로직을 분리하고 싶었습니다. 조회를 할 때는 조인이 발생하게 되는데, 조인이 발생하면 하나의 엔티티가 아니라 서로 다른 두 엔티티가 섞이기 때문입니다. 순수 엔티티 로직만 처리하는 레포지토리와 이를 분리하고 싶어 레포지토리를 분리하게 되었습니다!

realsigridjin replied:
@devjun10 추후 JPA를 사용하실 때에는 별도의 `readOnlyRepository` 를 만들어서 상속받아 관리하면 유지보수에 편리할 것이라고 생각합니다. Data JDBC 여도 비슷하게 접근할 수 있을 것 같아요.

[참고할 만한 링크](https://www.baeldung.com/spring-data-read-only-repository)
[참고할 만한 링크 2](https://kimjongmo.github.io/jpa/read-only-repo)

sigridjineth reviewed (approved):
안녕하세요, 리뷰어 Sigrid입니다.
프로젝트 초반인데 첫 단추를 잘 끼운 것 같네요. 개인적으로 들었던 생각을 적었으니 한 번 살펴봐주세요.
그러면 즐거운 프로젝트 첫 주가 되시길 바랍니다. 감사합니다.

donggi-lee-bit commented:
> 안녕하세요, 리뷰어 시그리드입니다. 먼저 2주간 잘 부탁드리구요. 언제든지 궁금하신 점이 있으면 메시지나 PR을 자주 남겨주시면 디스커션과 저의 팔로업에도 도움이 됩니다.

리뷰 감사합니다 sigrid !! 내일 팀원과 같이 리뷰해주신 내용 함께 학습하여 댓글 달아보겠습니다 좋은 밤 되십시오~~~ 😆

donggi-lee-bit committed:
Feat: Order 도메인 클래스 추가 - #20 (#21)

- DB 의 order 테이블을 사용하기 위한 Order 클래스 추가

donggi-lee-bit committed:
Feat: OrderDeatil 도메인 클래스 추가 - #20 (#22)

- DB 의 orderdetail 테이블을 사용하기 위한 OrderDetail 클래스 추가

donggi-lee-bit committed:
[BE] Spring Rest Docs 기능 추가 (#23)

* Feat: Spring Rest Docs 의존성 추가 - #14

- Gradle 버전에 맞게 Spring Rest Docs 의존성 추가

* Feat: 개발 환경, 테스트 환경 분리 - #14

- 테스트 환경을 분리하여 H2 DB 연결

* Feat: Spring Rest Docs 기능 추가 (테스트한 더미 데이터 추가되지 않음) - #14

- 테스트를 통한 API 명세를 확인하기 위해 Spring Rest Docs 추가
- 테스트한 더미 데이터는 추가되지 않는 이슈를 갖고 있음

donggi-lee-bit commented:
> 깃 브랜치 관리 기법이 인상적이네요. 저 그림은 일일이 그리신 건가요?

Jun 이 작업해주셨습니다 !!! 🥹

donggi-lee-bit committed:
Refactor: Swagger Contact 의 url 주소 변경 - #24

- 슬랙 -> Github url 로 수정

donggi-lee-bit committed:
Refactor: schema.sql 컬럼 comment 추가 및 name 컬럼 글자 제한 수 변경 - #24

devjun10 commented:
> ### 스키마 관련
> * 김영한 선생님의 강의를 보면 `일대다 단방향 매핑` 이 좋지 않다고 합니다. 위에 올려주신 스키마도 `일대다 양방향 매핑` 이 되어 있네요. 일대다 단방향이 아닌, 양방향으로 매핑 전략을 가져간 것에 대한 본인만의 주장과 근거를 제시해주세요. 지금 바로 완전한 답변을 주지 않으셔도 좋고, 점진적으로 학습 과정 속에서 느끼시는 사항을 답변으로 남겨주셔도 좋습니다.
> * 우리의 현재 요구사항에서 `dish` 테이블이 `orderdetail` 테이블의 정보를 참조해야 하는 내용이 있나요?
> 
> ### 테이블 설계 관련
> * 테이블 설계에서 하나의 테이블을 두 개의 테이블로 분리하여 보관하려는 것 같은데요. 그렇게 한다면 이유는 무엇인가요? 만약 상품의 정보를 별도의 테이블로 빼서 관리한다고 하면, 해당 테이블을 별도로 뺀 것에 대한 정보를 한 번 상위 개념에서 묶어줄 수 있을 것이고, 그러면 1:1 매핑 관계가 성립할 수 있지 않을까요?
> * 지금은 DB 스키마 설계를 먼저 했고 이에 바탕해서 데이터를 넣어보시면서 작업하려는 것 같은데요. 이것도 분명 훌륭한 방법인데, 저는 개인적으로 객체 설계를 먼저 하는 방향을 좋아합니다. 이는 철저히 개인 취향이니 만큼 한 번 고민해보시면 좋을 것 같습니다.
> 
> ### 조인 관련
> * 글쎄요. 구체적으로 어떻게 칼럼이 뻥튀기 된다고 말씀하시는 지 궁금하네요. `*` 표시로 내려주셔서 그런 것 아닐까요?
> * 지금 `JOIN` 을 하고 계신데, `OUTER JOIN` 과 `INNER JOIN` 의 차이가 무엇인지 학습해보고 본인만의 언어로 기술해주세요.

우선 시간내서 리뷰 주셔서 진심으로 감사합니다. PR 내용을 통해 몰랐던 많은 부분을 알 수 있었습니다. 
다음 PR때 주셨던 질문사항에 대한 답변, PR 내용 반영해보겠습니다. 감사합니다 🙏
